tests/PHPStan/Analyser/nsrt/array_splice.php

```diff
@@ -21,41 +21,195 @@ function insertViaArraySplice(array $arr): void
+
+	/** @var array{17: 'foo', b: 'bar', 19: 'baz'} $arr */
+	array_splice($arr, -1, 1, ['hello']);
+	assertType('array{0: \'foo\', b: \'bar\', 1: \'hello\'}', $arr);
+
+	/** @var array{17: 'foo', b: 'bar', 19: 'baz'} $arr */
+	array_splice($arr, 0, null, ['hello']);
+	assertType('array{\'hello\'}', $arr);
+
+	/** @var array{17: 'foo', b: 'bar', 19: 'baz'} $arr */
+	array_splice($arr, 0);
+	assertType('array{}', $arr);
+}
+
+function constantArraysWithOptionalKeys(array $arr): void
+{
+	// array{a?: 0, b: 1, c: 2}
+
+	/**
+	 * @see https://3v4l.org/jrqoZ
+	 * @var array{a?: 0, b: 1, c: 2} $arr
+	 */
+	array_splice($arr, 0, 1, ['hello']);
+	assertType('array{0: \'hello\', b?: 1, c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/lbUJG
+	 * @var array{a?: 0, b: 1, c: 2} $arr
+	 */
+	array_splice($arr, 1, 1, ['hello']);
+	assertType('array{a?: 0, b?: 1, 0: \'hello\', c?: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/7uPmV
+	 * @var array{a?: 0, b: 1, c: 2} $arr
+	 */
+	array_splice($arr, -1, 0, ['hello']);
+	assertType('array{a?: 0, b: 1, 0: \'hello\', c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/hB8pG
+	 * @var array{a?: 0, b: 1, c: 2} $arr
+	 */
+	array_splice($arr, 0, -1, ['hello']);
+	assertType('array{0: \'hello\', b?: 1, c?: 2}', $arr); // Could be array{0: 'hello', c: 2}
+
+	// array{a: 0, b?: 1, c: 2}
+
+	/**
+	 * @see https://3v4l.org/TjfHT
+	 * @var array{a: 0, b?: 1, c: 2} $arr
+	 */
+	array_splice($arr, 0, 1, ['hello']);
+	assertType('array{0: \'hello\', b?: 1, c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/D8PSE
+	 * @var array{a: 0, b?: 1, c: 2} $arr
+	 */
+	array_splice($arr, 1, 1, ['hello']);
+	assertType('array{a: 0, 0: \'hello\', c?: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/8RfDs
+	 * @var array{a: 0, b?: 1, c: 2} $arr
+	 */
+	array_splice($arr, -1, 0, ['hello']);
+	assertType('array{a: 0, b?: 1, 0: \'hello\', c: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/sPfpN
+	 * @var array{a: 0, b?: 1, c: 2} $arr
+	 */
+	array_splice($arr, 0, -1, ['hello']);
+	assertType('array{0: \'hello\', c?: 2}', $arr); // Could be array{0: 'hello', c: 2}
+
+	// array{a: 0, b: 1, c?: 2}
+
+	/**
+	 * @see https://3v4l.org/Ddpku
+	 * @var array{a: 0, b: 1, c?: 2} $arr
+	 */
+	array_splice($arr, 0, 1, ['hello']);
+	assertType('array{0: \'hello\', b: 1, c?: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/O4LLi
+	 * @var array{a: 0, b: 1, c?: 2} $arr
+	 */
+	array_splice($arr, 1, 1, ['hello']);
+	assertType('array{a: 0, 0: \'hello\', c?: 2}', $arr);
+
+	/**
+	 * @see https://3v4l.org/QQO84
+	 * @var array{a: 0, b: 1, c?: 2} $arr
+	 */
+	array_splice($arr, -1, 0, ['hello']);
+	assertType('array{a?: 0, 0: \'hello\', b?: 1, c?: 2}', $arr); // Could be array{a?: 0, 0: 'hello', b: 1, c?: 2}
+
+	/**
+	 * @see https://3v4l.org/K5RDp
+	 * @var array{a: 0, b: 1, c?: 2} $arr
+	 */
+	array_splice($arr, 0, -1, ['hello']);
+	assertType('array{0: \'hello\', c?: 2}', $arr); // Could be array{0: 'hello', b?: 1, c?: 2}
```

Comment:
Some types with either negative offsets or negative limits are not optimal but this one is wrong. Not sure what to do here tbh., the handling of those is extremely hard in combination with optional keys. Without optional keys my current approach seems to work fine.

Ideas about this

- Keep and see if it causes any issues
- Degrade to general array if negative offsets/limits are involved in combination with optional keys
- Rewrite and do something semi-crazy with maybe `getFiniteTypes()` and avoid optional keys completely but have some constant array unions instead?

Reply:
I would prefer a general array over a wrong type (which can lead to more false-positives along the line)

Reply:
Thx for the input! I think I know one of the cause, will take another look later again.

Reply:
Ok, so I could add support for negative offsets with optional keys, but not for negative lengths with optional keys. I'm explicitly degrading for the latter case too now, should be safer indeed..

Reply:
found unfortunately a problematic case for negative offsets too. those are both excluded now..

Reply:
sorry for the back and forth, but I found a better way using `ConstantArrayType::getAllArrays()` that leads to perfect accuracy and only degrades if the arrays get too complex.